documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,652 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+}
+
+public partial class Gamepads : IReadOnlyList<IGamepad>
+{
+    public event Action<ButtonEvent<IGamepad, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<IGamepad, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<IGamepad, Vector2>>? ThumbstickMove;
+    public event Action<AxisEvent<IGamepad, float>>? TriggerMove;
+}
+
+public partial class Joysticks : IReadOnlyList<IJoystick>
+{
+    public event Action<ButtonEvent<IJoystick, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<IJoystick, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<IJoystick, float>>? AxisMove;
+    public event Action<AxisEvent<IJoystick, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(TDevice Device, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(TDevice Device, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(TDevice Device, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(TDevice Device, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse.
```

Comment:
We should probably clarify that this won't (immediately) change `IMouse.State`, or at least not without an `Update()`

Reply:
Clarified.